apps/src/StudioApp.js

```diff
@@ -843,17 +868,96 @@ StudioApp.prototype.createModalDialog = function(options) {
   return this.feedback_.createModalDialog(options);
 };
 
+StudioApp.prototype.recordUserViewedHint_ = function (hint, config) {
+  hint.alreadySeen = true;
+
+  authoredHintUtils.recordUnfinishedHint({
+    // level info
+    script_id: config.scriptId,
+    level_id: config.level.scriptLevelId,
+
+    // hint info
+    hint_id: hint.hintId,
+    hint_class: hint.hintClass,
+    hint_type: hint.hintType,
+  });
+};
+
+StudioApp.prototype.onReportComplete = function (response) {
+  authoredHintUtils.finishHints({
+    time: ((new Date().getTime()) - this.initTime),
+    attempt: this.attempts,
+    test_result: this.lastTestResult,
+    activity_id: response && response.activity_id,
+    level_source_id: response && response.level_source_id,
+  });
+};
+
+StudioApp.prototype.showHint_ = function (hint, config) {
+
+  $('#prompt-icon').qtip({
+    content: {
+      text: function(html, api) {
+        var container = document.createElement('div');
+        container.innerHTML = require('./templates/hintSelect.html.ejs')();
+
+        var instructionsButton = container.querySelector('#inst');
+        dom.addClickTouchEvent(instructionsButton, function () {
+          this.showInstructions_(config.level, false);
+          api.destroy();
+        }.bind(this));
+
+        var hintButton = container.querySelector('#newhint');
+        dom.addClickTouchEvent(hintButton, function () {
+          api.set('content.text', hint.content);
+          this.recordUserViewedHint_(hint, config);
+        }.bind(this));
+
+        return container;
+      }.bind(this),
+      title: {
+        button: $('<div class="tooltip-x-close"/>')
+      }
+    },
+    style: {
+      classes: "cdo-qtips",
+      tip: {
+        width: 20,
+        height: 20
+      }
+    },
+    position: {
+      my: "bottom left",
+      at: "top right"
+    },
+    hide: {
+      event: 'click mousedown touchstart unfocus'
+    },
+    show: false // don't show on mouseover
+  }).qtip('show');
+
+};
+
```

Comment:
Yeah, there's lots of opportunity to make things better here. If we're saying that we will spend some time doing tech debt type things during hour of code, this would be a great task for someone to jump on (or multiple someones to pair on).